File: backtester/backend/schemas/strategy.py
"""Strategy API Schemas."""

from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class StrategyListResponse(BaseModel):
    """전략 목록 응답"""
    success: bool = True
    data: List[Dict[str, Any]]
    total: int


class StrategyDetailResponse(BaseModel):
    """전략 상세 응답"""
    success: bool = True
    data: Dict[str, Any]


class StrategyBuildRequest(BaseModel):
    """전략 빌드 요청"""
    id: str = Field(..., description="전략 ID")
    name: str = Field(..., description="전략 이름")
    description: Optional[str] = Field(None, description="전략 설명")
    category: str = Field(default="custom", description="카테고리")
    indicators: List[Dict[str, Any]] = Field(..., description="지표 목록")
    entry: Dict[str, Any] = Field(..., description="진입 조건")
    exit: Dict[str, Any] = Field(..., description="청산 조건")
    params: Optional[Dict[str, Dict[str, Any]]] = Field(None, description="파라미터")
    risk_management: Optional[Dict[str, Any]] = Field(None, description="리스크 관리")


class StrategyBuildResponse(BaseModel):
    """전략 빌드 응답"""
    success: bool = True
    data: Dict[str, Any]
    message: Optional[str] = None

File: backtester/frontend/src/types/strategy.ts
/**
 * 전략 타입 정의
 */

/**
 * 파라미터 정의 (슬라이더용)
 */
export interface ParamDefinition {
  default: number;
  min: number;
  max: number;
  step?: number;
  type?: "int" | "float";
  label?: string;
}

export interface Strategy {
  id: string;
  name: string;
  description: string;
  category: string;
  tags?: string[];
  params?: Record<string, ParamDefinition>;  // 파라미터 정의
}

export interface StrategyDetail extends Strategy {
  indicators: IndicatorConfig[];
  entry: ConditionGroup;
  exit: ConditionGroup;
  param_values: Record<string, number>;
  risk_management: RiskConfig;
}

export interface IndicatorConfig {
  id: string;
  alias?: string;
  params: Record<string, number>;
}

export interface Condition {
  indicator: string;
  operator: string;
  compare_to: string | number;
}

export interface ConditionGroup {
  logic: "AND" | "OR";
  conditions: Condition[];
}

export interface RiskConfig {
  stop_loss?: { enabled: boolean; percent: number };
  take_profit?: { enabled: boolean; percent: number };
  trailing_stop?: { enabled: boolean; percent: number };
}

export interface StrategyListResponse {
  success: boolean;
  data: Strategy[];
  total: number;
}

export interface StrategyDetailResponse {
  success: boolean;
  data: StrategyDetail;
}

export interface Category {
  id: string;
  name: string;
}

export interface CategoryListResponse {
  success: boolean;
  data: Category[];
}
